### PR TITLE
Add skeletons for Mizar GUI

### PR DESCRIPTION
**Pull request by akopich**
This adds some skeletons of Widgets needed for that part
of Mizar GUI dealing with sampling with frametrack comparison.

Bug http://b/237284844
Test Compile, Manual

### CODE REVIEW

**Comment by akopich** [comment]
https://screenshot.googleplex.com/7UPY65oQyaw3G9a.png

**Comment by akopich** [comment]
screenshot/7UPY65oQyaw3G9a.png

File: src/Mizar/MizarMainWindow.ui [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
+ <class>MainWindow</class>
+ <widget class="QMainWindow" name="MainWindow">
```
Comment (florian-kuebler) [inline src/Mizar/MizarMainWindow.ui:4]: Please be consistent with the "name" property. Either use `snake_case_`, `camelCase`, or `PascalCase`.

Comment (florian-kuebler) [reply src/Mizar/MizarMainWindow.ui:4]: Same for the other ui file.

File: src/Mizar/MizarMainWindow.ui [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    <y>0</y>
+    <width>986</width>
+    <height>449</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>MainWindow</string>
+  </property>
+  <widget class="QWidget" name="centralwidget">
+   <layout class="QVBoxLayout" name="verticalLayout">
```
Comment (florian-kuebler) [inline src/Mizar/MizarMainWindow.ui:17]: you might want to consider having more "concrete" names for the vertical and horizontal layouts. The tree nature of the UI elements won't be easily reflected in the generated class. You'll end up with `main_window.ui->verticalLayout`, `main_window.ui->verticalLayout_2`, or `main_window.ui->horizontalLayout_2`. It'll be hard to know which one corresponds to what layout.

File: src/Mizar/SamplingWithFrameTrackInputWidget.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#define MIZAR_SAMPLING_WITH_FRAME_TRACK_INPUT_WIDGET_H_
+
+#include <QWidget>
+#include <memory>
+#include <string_view>
+
+#include "MizarData/MizarPairedData.h"
+
+namespace Ui {
+class SamplingWithFrameTrackInputWidget;
```
Comment (florian-kuebler) [inline src/Mizar/SamplingWithFrameTrackInputWidget.h:15]: I am probably missing some Qt specifics for custom Widgets, but this forward declaration here looks strange.

File: src/Mizar/SamplingWithFrameTrackInputWidget.ui [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+       <item>
+        <widget class="QLabel" name="label_2">
+         <property name="maximumSize">
+          <size>
+           <width>70</width>
+           <height>16777215</height>
+          </size>
+         </property>
+         <property name="text">
+          <string>From ms:</string>
```
Comment (florian-kuebler) [inline src/Mizar/SamplingWithFrameTrackInputWidget.ui:49]: That is a bit unclear to me. Is this what ms timestamp of the capture we should you as start for the comparison?

File: src/Mizar/SamplingWithFrameTrackInputWidget.h (hunk 3 above)
Comment (beckerhe) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:15]: It's a requirement for the pimpl idiom.

File: src/Mizar/SamplingWithFrameTrackInputWidget.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+class SamplingWithFrameTrackInputWidgetTmpl : public QWidget {
+ public:
+  explicit SamplingWithFrameTrackInputWidgetTmpl(QWidget* parent);
+  ~SamplingWithFrameTrackInputWidgetTmpl() override;
+
+  void Init(const PairedData* data, const QString& name);
+
+ private:
+  std::unique_ptr<Ui::SamplingWithFrameTrackInputWidget> ui_;
+};
```
Comment (beckerhe) [inline src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: Classes that derive from `QObject` shouldn't/can't be templated because this is not supported by the MOC preprocessor.

Couldn't you just template `Init`?

Comment (beckerhe) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: If not I would suggest to not give this widget direct access to `PairedData` but rather define an interface - similar to what Qt does with `QAbstractItemModel`.

File: src/Mizar/SamplingWithFrameTrackInputWidget.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace orbit_mizar {
+
+template <typename PairedData>
+class SamplingWithFrameTrackInputWidgetTmpl : public QWidget {
+ public:
+  explicit SamplingWithFrameTrackInputWidgetTmpl(QWidget* parent);
+  ~SamplingWithFrameTrackInputWidgetTmpl() override;
+
+  void Init(const PairedData* data, const QString& name);
```
Comment (beckerhe) [inline src/Mizar/SamplingWithFrameTrackInputWidget.h:26]: What about giving that function a more descriptive name. Something like `SetSamplingData` or so?

File: src/Mizar/SamplingWithFrameTrackInputWidget.h (hunk 5 above)
Comment (beckerhe) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: (Same goes for the other class)

File: src/Mizar/SamplingWithFrameTrackInputWidget.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef MIZAR_SAMPLING_WITH_FRAME_TRACK_INPUT_WIDGET_H_
```
Comment (beckerhe) [inline src/Mizar/SamplingWithFrameTrackInputWidget.h:5]: I think both Widget classes should be moved into a separate module - maybe `MizarWidgets` or something else to your liking. The reason is that otherwise they can't be tested.

File: src/Mizar/CMakeLists.txt [hunk 8]
```diff
@@ -4,22 +4,38 @@
...
+target_include_directories(Mizar PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+target_include_directories(Mizar PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 
-target_link_libraries(Mizar 
-PRIVATE CaptureClient
+target_link_libraries(Mizar
+        PUBLIC QtUtils
+        Qt5::Widgets
+        Qt5::Core
+        PRIVATE CaptureClient
```
Comment (beckerhe) [inline src/Mizar/CMakeLists.txt:31]: Since Mizar is an executable target, there is no difference between PRIVATE and PUBLIC. Imho all dependencies should be marked as PRIVATE

File: src/Mizar/CMakeLists.txt [hunk 9]
```diff
@@ -4,22 +4,38 @@
...
+        SamplingWithFrameTrackInputWidget.h
+        SamplingWithFrameTrackInputWidget.cpp
+        SamplingWithFrameTrackInputWidget.ui
+        SamplingWithFrameTrackWidget.h
+        SamplingWithFrameTrackWidget.cpp
+        SamplingWithFrameTrackWidget.ui)
 
-target_include_directories(Mizar PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}) 
+target_include_directories(Mizar PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+target_include_directories(Mizar PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
```
Comment (beckerhe) [inline src/Mizar/CMakeLists.txt:25]: Why this change?

File: src/Mizar/MizarMainWindow.cpp [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+#include "Mizar/ui_MizarMainWindow.h"
+
+namespace orbit_mizar {
+MizarMainWindow::MizarMainWindow() = default;
+MizarMainWindow::~MizarMainWindow() = default;
+
+void MizarMainWindow::Setup() {
+  ui_ = std::make_unique<Ui::MainWindow>();
+  ui_->setupUi(this);
+}
```
Comment (beckerhe) [inline src/Mizar/MizarMainWindow.cpp:18]: This should be done in the constructor.

File: src/Mizar/MizarMainWindow.ui (hunk 1 above)
Comment (akopich) [reply src/Mizar/MizarMainWindow.ui:4]: Discussed offline. Agreed to keep the widget names PascalCase and make everything else snake_case_

File: src/Mizar/SamplingWithFrameTrackInputWidget.ui (hunk 4 above)
Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.ui:49]: Exactly. 
I plan to remove the element as soon as orbit itself can handle this.

File: src/Mizar/SamplingWithFrameTrackInputWidget.h (hunk 5 above)
Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: > Couldn't you just template Init?

I can try. But this way I will not be able to store a pointer to PairedData instance in the widget. 

>(Same goes for the other class)

Sorry, which class? I think, it's the only template class I add in the PR.

File: src/Mizar/SamplingWithFrameTrackInputWidget.ui (hunk 4 above)
Comment (florian-kuebler) [reply src/Mizar/SamplingWithFrameTrackInputWidget.ui:49]: Can we find a slightly more descriptive name? "Comparison start (ms):" maybe?

File: src/Mizar/SamplingWithFrameTrackInputWidget.h (hunk 5 above)
Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: `Function template partial specialization is not allowed`
... I guess, I can't.

Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:30]: Done. The method is templated, the class isn't.

File: src/Mizar/SamplingWithFrameTrackInputWidget.h (hunk 6 above)
Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.h:26]: `Init` name makes the user think "unless I call the method, an instance of the class is useless", `SetWhatever` -- not so much. Also, the method signature clearly states what exactly we set, doesn't it?

File: src/Mizar/CMakeLists.txt (hunk 9 above)
Comment (akopich) [reply src/Mizar/CMakeLists.txt:25]: For absolutely no good reason at all. Rolling it back.

Comment (akopich) [reply src/Mizar/CMakeLists.txt:25]: For absolutely no good reason at all. Rolling it back.

File: src/Mizar/SamplingWithFrameTrackInputWidget.ui (hunk 4 above)
Comment (akopich) [reply src/Mizar/SamplingWithFrameTrackInputWidget.ui:49]: I can replace "From ms" with "Start (ms)", but repeating Baseline/Comparison I find to be redundant.

**Comment by antonrohr** [comment]
Just a heads up, I don't think I will find time to do this review today.

File: src/MizarWidgets/MizarMainWindow.ui [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    <y>0</y>
+    <width>986</width>
+    <height>449</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Mizar</string>
+  </property>
+  <widget class="QWidget" name="central_widget_">
+   <layout class="QVBoxLayout" name="verticalLayout">
```
Comment (florian-kuebler) [inline src/MizarWidgets/MizarMainWindow.ui:17]:
```suggestion
   <layout class="QVBoxLayout" name="vertical_layout_">
```

File: src/MizarWidgets/MizarMainWindow.ui [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      <property name="currentIndex">
+       <number>0</number>
+      </property>
+      <widget class="QWidget" name="sampling_with_frametrack_tab_">
+       <attribute name="title">
+        <string>Sampling with Frame Track</string>
+       </attribute>
+       <layout class="QVBoxLayout" name="verticalLayout_2">
+        <item>
+         <layout class="QHBoxLayout" name="layout">
```
Comment (florian-kuebler) [inline src/MizarWidgets/MizarMainWindow.ui:29]: sampling_with_frame_track_h_box_layout_

File: src/MizarWidgets/MizarMainWindow.ui [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      <widget class="QWidget" name="tab_2">
+       <attribute name="title">
+        <string>Not implemented yet</string>
+       </attribute>
+      </widget>
+     </widget>
+    </item>
+   </layout>
+  </widget>
+  <widget class="QStatusBar" name="statusbar"/>
```
Comment (florian-kuebler) [inline src/MizarWidgets/MizarMainWindow.ui:46]: status_bar_

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+       <item>
+        <widget class="QLabel" name="label_2">
+         <property name="maximumSize">
+          <size>
+           <width>70</width>
+           <height>16777215</height>
+          </size>
+         </property>
+         <property name="text">
+          <string>From ms:</string>
```
Comment (florian-kuebler) [inline src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:49]: Still not 100% sure, if this is descriptive enough. But maybe its fine.

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+         <property name="maximumSize">
+          <size>
+           <width>100</width>
+           <height>16777215</height>
+          </size>
+         </property>
+        </widget>
+       </item>
+       <item>
+        <widget class="QLabel" name="label_3">
```
Comment (florian-kuebler) [inline src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:70]: frame_track_label

File: src/MizarWidgets/SamplingWithFrameTrackWidget.ui [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    <x>0</x>
+    <y>0</y>
+    <width>823</width>
+    <height>456</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Form</string>
+  </property>
+  <layout class="QVBoxLayout" name="verticalLayout">
```
Comment (florian-kuebler) [inline src/MizarWidgets/SamplingWithFrameTrackWidget.ui:16]: vertical_layout_

File: src/MizarWidgets/MizarMainWindow.ui (hunk 12 above)
Comment (akopich) [reply src/MizarWidgets/MizarMainWindow.ui:29]: But it's also vertical layout. Just like the one you suggest to call sampling_with_frame_track_vertical_layout_

Comment (akopich) [reply src/MizarWidgets/MizarMainWindow.ui:29]: ... NVM, I've just removed the "layout" and the item it was wrapped in.

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    <x>0</x>
+    <y>0</y>
+    <width>569</width>
+    <height>233</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Form</string>
+  </property>
+  <layout class="QGridLayout" name="gridLayout">
```
[the remark this answers is not in the document]
Comment (akopich) [reply src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:16]: This layout wasn't even shown in Qt Designer. Thanks.

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui (hunk 14 above)
Comment (akopich) [reply src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:49]: Tooltips are good. But not in this PR.

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui (hunk 15 above)
Comment (akopich) [reply src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:70]: frame_track_label_ ?

File: src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui (hunk 14 above)
Comment (florian-kuebler) [reply src/MizarWidgets/SamplingWithFrameTrackInputWidget.ui:49]: Would be fine with me. But note that you can simply add a tooltip in the QtDesigner.

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks

**Review by beckerhe — APPROVED** [review]
Thank you for moving all the code around. LGTM!

File: src/MizarWidgets/MizarMainWindow.cpp [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+#include "MizarWidgets/MizarMainWindow.h"
+
+#include <memory>
+
+#include "ui_MizarMainWindow.h"
+
+namespace orbit_mizar_widgets {
+
+MizarMainWindow::MizarMainWindow() {
+  ui_ = std::make_unique<Ui::MainWindow>();
```
Comment (antonrohr) [inline src/MizarWidgets/MizarMainWindow.cpp:14]: I personally have a preference for: `
MizarMainWindow::MizarMainWindow(): ui_(std::make_unique<...>) {`
feel free to ignore

**Review by antonrohr — APPROVED** [review]
LGTM!
